Utilities/O2Device/include/O2Device/Utilities.h

```diff
@@ -64,6 +77,49 @@ bool AddDataBlock(O2Message& parts, o2::header::Stack&& inputStack, ContainerT&&
   return true;
 }
 
+namespace internal
+{
+
+template <typename I, typename F>
+auto forEach(I begin, I end, F function)
+{
+  using span = gsl::span<const byte>;
+  using gsl::narrow_cast;
+  for (auto it = begin; it != end; ++it) {
+    byte* headerBuffer{ nullptr };
+    span::index_type headerBufferSize{ 0 };
+    if (*it != nullptr) {
+      headerBuffer = reinterpret_cast<byte*>((*it)->GetData());
+      headerBufferSize = narrow_cast<span::index_type>((*it)->GetSize());
+    }
+    ++it;
+    byte* dataBuffer{ nullptr };
+    span::index_type dataBufferSize{ 0 };
+    if (*it != nullptr) {
+      dataBuffer = reinterpret_cast<byte*>((*it)->GetData());
+      dataBufferSize = narrow_cast<span::index_type>((*it)->GetSize());
+    }
+
+    // call the user provided function
+    function(span{ headerBuffer, headerBufferSize }, span{ dataBuffer, dataBufferSize });
+  }
+  return std::move(function);
```

Comment:
Trying to understand why the function object is returned by move. Taking as an example:
```
O2message message1;
O2message message2;
forEach(message1, forEach(message2, funtion));
```
the function is applied first to message2 and then the same to message1. But one would probably write it in a different way.

Reply:
I took inspiration from http://en.cppreference.com/w/cpp/algorithm/for_each , I think this indeed let's you chain algorithms without the need to instantiate/copy the possibly heavy functors...